I tested this locally by checking out f60382bf21601b1c6708d094fadb13de59a77278 and applying this patch. This was the gradle call (using the test catalog from f60382bf21601b1c6708d094fadb13de59a77278's CI build)

```
GITHUB_ACTIONS=1 ./gradlew --info --continue --no-scan \
    -PmaxTestRetries=1 -PmaxTestRetryFailures=3 \
    -PmaxQuarantineTestRetries=3 -PmaxQuarantineTestRetryFailures=0 \
    -Pkafka.test.catalog.file=/Users/darthur/Downloads/combined-test-catalog.txt \
    :clients:quarantinedTest --rerun-tasks
```


```
> Task :clients:quarantinedTest FAILED

ConsumerHeartbeatRequestManagerTest > testUnsupportedVersion() STANDARD_OUT
    [2024-12-05 16:36:37,437] ERROR ConsumerGroupHeartbeatRequest failed due to UNSUPPORTED_VERSION: null (org.apache.kafka.clients.consumer.internals.ConsumerHeartbeatRequestManager:411)
org.apache.kafka.clients.consumer.internals.ConsumerHeartbeatRequestManagerTest.testUnsupportedVersion() failed, log available in /Users/darthur/Code/Apache/kafka/clients/build/reports/testOutput/org.apache.kafka.clients.consumer.internals.ConsumerHeartbeatRequestManagerTest.testUnsupportedVersion().test.stdout

ConsumerHeartbeatRequestManagerTest > testUnsupportedVersion() FAILED
    org.opentest4j.AssertionFailedError: expected: <The cluster does not support the new consumer group protocol. Set group.protocol=classic on the consumer configs to revert to the classic protocol until the cluster is upgraded.> but was: <The cluster does not support the new CONSUMER group protocol. Set group.protocol=classic on the consumer configs to revert to the CLASSIC protocol until the cluster is upgraded.>
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
        at app//org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
        at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
        at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
        at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1145)
        at app//org.apache.kafka.clients.consumer.internals.ConsumerHeartbeatRequestManagerTest.testUnsupportedVersion(ConsumerHeartbeatRequestManagerTest.java:616)

19 tests completed, 4 failed, 3 skipped
Finished generating test XML results (0.01 secs) into: /Users/darthur/Code/Apache/kafka/clients/build/test-results/quarantinedTest
Generating HTML test report...
Finished generating test html results (0.005 secs) into: /Users/darthur/Code/Apache/kafka/clients/build/reports/tests/quarantinedTest
There were failing tests. See the report at: file:///Users/darthur/Code/Apache/kafka/clients/build/reports/tests/quarantinedTest/index.html
Copy JUnit XML for clients to /Users/darthur/Code/Apache/kafka/build/junit-xml/clients/quarantinedTest
[ant:copy] Copying 11 files to /Users/darthur/Code/Apache/kafka/build/junit-xml/clients/quarantinedTest

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/darthur/Code/Apache/kafka/build.gradle' line: 613

* What went wrong:
Execution failed for task ':clients:quarantinedTest'.
> Failing this task since 'clients:quarantinedTest' had test failures.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --debug option to get more log output.
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.10.2/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 54s
12 actionable tasks: 12 executed
Watched directory hierarchies: [/Users/darthur/Code/Apache/kafka]
```

So we see the test fail, but we now see the XML files copied to the right place.
```
[ant:copy] Copying 11 files to /Users/darthur/Code/Apache/kafka/build/junit-xml/clients/quarantinedTest
```

--- a/build.gradle
+++ b/build.gradle
@@ -546,6 +546,7 @@ subprojects {
   task quarantinedTest(type: Test, dependsOn: compileJava) {
     ext {
       isGithubActions = System.getenv('GITHUB_ACTIONS') != null
+      hadFailure = false  // Used to track if any tests failed, see afterSuite below
     }
 
     // Disable caching and up-to-date for this task. We always want quarantined tests
@@ -584,6 +585,13 @@ subprojects {
       }
     }
 
+    // As we process results, check if there were any test failures.
+    afterSuite { desc, result ->
+      if (result.resultType == TestResult.ResultType.FAILURE) {
+        ext.hadFailure = true
+      }
+    }
+
     // This closure will copy JUnit XML files out of the sub-project's build directory and into
     // a top-level build/junit-xml directory. This is necessary to avoid reporting on tests which
     // were not run, but instead were restored via FROM-CACHE. See KAFKA-17479 for more details.
@@ -596,6 +604,11 @@ subprojects {
           ant.fileset(dir: "${quarantinedTest.reports.junitXml.entryPoint}") {
             ant.include(name: "**/*.xml")
           }
+        }
+        // If there were any test failures, we want to fail the task to prevent the failures
+        // from being cached.
+        if (ext.hadFailure) {
+          throw new GradleException("Failing this task since '${project.name}:${name}' had test failures.")
         }
       }
     }

--- a/build.gradle
+++ b/build.gradle
@@ -546,7 +546,6 @@ subprojects {
   task quarantinedTest(type: Test, dependsOn: compileJava) {
     ext {
       isGithubActions = System.getenv('GITHUB_ACTIONS') != null
-      hadFailure = false  // Used to track if any tests failed, see afterSuite below
     }
 
     // Disable caching and up-to-date for this task. We always want quarantined tests
@@ -585,13 +584,6 @@ subprojects {
       }
     }
 
-    // As we process results, check if there were any test failures.
-    afterSuite { desc, result ->
-      if (result.resultType == TestResult.ResultType.FAILURE) {
-        ext.hadFailure = true
-      }
-    }
-
     // This closure will copy JUnit XML files out of the sub-project's build directory and into
     // a top-level build/junit-xml directory. This is necessary to avoid reporting on tests which
     // were not run, but instead were restored via FROM-CACHE. See KAFKA-17479 for more details.
@@ -604,11 +596,6 @@ subprojects {
           ant.fileset(dir: "${quarantinedTest.reports.junitXml.entryPoint}") {
             ant.include(name: "**/*.xml")
           }
-        }
-        // If there were any test failures, we want to fail the task to prevent the failures
-        // from being cached.
-        if (ext.hadFailure) {
-          throw new GradleException("Failing this task since '${project.name}:${name}' had test failures.")
         }
       }
     }

